### PR TITLE
feat(scheduler): add automatic scheduler switching feature

### PR DESCRIPTION
**Pull request by PoloDyBala**
## Description

To address potential issues with Scheduler failures during P2P downloads, this project aims to implement an automatic switching feature at the Peer node. When a Scheduler encounters an issue and cannot properly handle downloads, the Peer node will switch to other Scheduler nodes using a hash ring to continue the download, thereby preventing download failures and improving success rates.

## Requirements

1. Scheduler Availability Check During Peer Registration: Ensure that the Scheduler is available during the Peer node registration process.

2. Automatic Scheduler Switching: If the current Scheduler is unavailable, automatically switch to another Scheduler node until a usable Scheduler is found.

3. Feature Development: Complete the development of the automatic switching logic and Scheduler availability detection.
Testing: Implement unit tests and end-to-end (E2E) tests to ensure the reliability and stability of the feature.
## Related Issue

- [https://github.com/dragonflyoss/client/issues/551](url)

## Motivation and Context

Please refer to the following document link for detailed instructions

- [https://www.yuque.com/baimo/fv9qk8/ghfsq8uldrer2p7r](url)

1. Location of corresponding configuration items：`dragonfly-client-config/src/dfdaemon.rs`
```   
    // cooldown_duration is the scheduler's cooldown time (in seconds).
    // When the scheduler becomes unavailable due to an exception, this field specifies
    // the time period during which the scheduler will not be accessed.
    #[serde(
        default = "default_scheduler_cooldown_interval",
        with = "humantime_serde"
    )]
    pub cooldown_interval: Duration,

    // max_attempts is the maximum number of attempts to connect to the scheduler.
    // If the connection to the scheduler fails, dfdaemon will retry up to this number of times.
    #[serde(default = "default_scheduler_max_attempts")]
    #[validate(range(min = 1))]
    pub max_attempts: u32,

    // refresh_threshold is the threshold for refreshing available scheduler nodes.
    // When the number of attempts to connect to the scheduler reaches this value,
    // dfdaemon will refresh the available scheduler nodes.
    #[serde(default = "default_scheduler_refresh_threshold")]
    #[validate(range(min = 1))]
    pub refresh_threshold: u32,
```
Simultaneously configure the corresponding default value function：
```
// default_scheduler_cooldown_interval is the default cooldown time (in seconds) for the scheduler.
// This duration specifies the time period during which the scheduler will not be accessed
// after it becomes unavailable due to an exception.
#[inline]
fn default_scheduler_cooldown_interval() -> Duration {
    Duration::from_secs(60)
}

// default_scheduler_max_attempts is the default maximum number of attempts to connect to the scheduler.
// If the connection to the scheduler fails, dfdaemon will retry up to this number of times.
#[inline]
fn default_scheduler_max_attempts() -> u32 {
    3
}

// default_scheduler_refresh_threshold is the default threshold for refreshing available scheduler nodes.
// When the number of attempts to connect to the scheduler reaches this value,
// dfdaemon will refresh the available scheduler nodes.
#[inline]
fn default_scheduler_refresh_threshold() -> u32 {
    10
}
```
2. Add error type: `dragonfly-client-core/src/error/mod.rs`
```
    // ExceededMaxAttempts is the error when the maximum number of attempts is 
   exceeded.
    #[error("exceeded maximum number of attempts")]
    ExceededMaxAttempts,

    // SchedulerNotServing is the error when the current scheduler is not serving.
    #[error("current scheduler is not serving")]
    SchedulerNotServing,
```

3. Add switch logic and scheduler health check：`dragonfly-client/src/grpc/scheduler.rs`
4. The `unavailable_scheduler_addrs` field is only valid between two refresh operations, so in order to maintain synchronization between the `available_scheduler_addrs` and `unavailable_scheduler_addrs` fields, a cleaning logic has been added to the `update_available_scheduler_addrs` function：`dragonfly-client/src/grpc/scheduler.rs`

```
    // Clean the unavailable scheduler addrs.
    let mut unavailable_scheduler_addrs = 
self.unavailable_scheduler_addrs.write().await;
    unavailable_scheduler_addrs.clear();
    drop(unavailable_scheduler_addrs);
```

### Step 1: Open multiple Scheduler containers with different IPs

image

### Step 2: Terminate some Scheduler containers within two Refresh intervals
Now you can see that there is only one Scheduler container：
image

### Step3: Run the local client and download the file

image

### Step 4: View log

From the log in the following figure, it can be seen that first, the Scheduler was selected based on task_id, but the connection failed because the corresponding container was terminated.
Then the error log below contains information about failed attempts to connect to other schedulers, ultimately selecting the only remaining Scheduler container
image

## Concurrency testing
1. `concurrent_dfget_test.sh`
```
URLS_TO_DOWNLOAD=(
  "https://mirrors.huaweicloud.com/nodejs/v0.1.100/SHASUMS256.txt"
  "https://mirrors.huaweicloud.com/nodejs/v0.1.101/SHASUMS256.txt"
  "https://mirrors.huaweicloud.com/nodejs/v0.1.102/SHASUMS256.txt"
  "https://mirrors.huaweicloud.com/nodejs/v0.1.103/SHASUMS256.txt"
  "https://mirrors.huaweicloud.com/nodejs/v0.1.104/SHASUMS256.txt"
  "https://mirrors.huaweicloud.com/nodejs/v0.1.14/SHASUMS256.txt"
  "https://mirrors.huaweicloud.com/nodejs/v0.1.15/SHASUMS256.txt"
  "https://mirrors.huaweicloud.com/nodejs/v0.1.16/SHASUMS256.txt"
  "https://mirrors.huaweicloud.com/nodejs/v0.1.17/SHASUMS256.txt"
  "https://mirrors.huaweicloud.com/nodejs/v0.1.18/SHASUMS256.txt"
)
OUTPUT_DIR="../download"

mkdir -p ${OUTPUT_DIR}

for i in "${!URLS_TO_DOWNLOAD[@]}"; do
  URL=${URLS_TO_DOWNLOAD[$i]}
  OUTPUT_FILE="${OUTPUT_DIR}/file_$((i+1)).txt"
  LOG_FILE="${OUTPUT_DIR}/file_$((i+1)).log"
  echo "Starting download for ${URL}..."
  cargo run --bin dfget -- -O ${OUTPUT_FILE} ${URL} > ${LOG_FILE} 2>&1 &
done

wait
echo "All downloads completed."
```
2. Result
屏幕截图 2024-07-29 101337

## Unit test
https://n0d5zorb70j.feishu.cn/docx/LDnKdfA4Wo3y34x27YXcz9HVnse?from=from_copylink

### CODE REVIEW

File: dragonfly-client/src/grpc/scheduler.rs [hunk 1]
```diff
@@ -434,36 +442,139 @@ impl SchedulerClient {
...
+        let cooldown_interval = config.scheduler.cooldown_interval;
+        let max_attempts = config.scheduler.max_attempts;
+        let refresh_threshold = config.scheduler.refresh_threshold;
+        let mut attempts = 0;
+
+        // Traverse through available scheduler collections.
+        let hashring = self.hashring.read().await;
+        for vnode in hashring.clone().into_iter() {
+            let scheduler_addr = vnode.addr;
+            let unavailable_scheduler_addrs = self.unavailable_scheduler_addrs.read().await;
```
Comment (xujihui1985) [inline dragonfly-client/src/grpc/scheduler.rs:491]: there is a potential issue with acquiring both read and write locks on unavailable_scheduler_addrs without properly releasing the read lock first, and do we need async lock here? I don't see a sign the lock is cross the await boundary

**Comment by xujihui1985** [comment]
after drop the unavailable_scheduler_addrs, is there any mechanism that the addr will be add back when the scheduler instance work normally?

File: dragonfly-client/src/grpc/scheduler.rs (hunk 1 above)
Comment (PoloDyBala) [reply dragonfly-client/src/grpc/scheduler.rs:491]: resolved

**Comment by PoloDyBala** [comment]
> after drop the unavailable_scheduler_addrs, is there any mechanism that the addr will be add back when the scheduler instance work normally?

The design intent is for `unavailable_scheduler_addrs` to be valid only during the interval between two refreshes. Therefore, when `available_scheduler_addrs` is updated, the corresponding `unavailable_scheduler_addrs` will be released.

File: dragonfly-client/src/grpc/health.rs [hunk 2]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-use dragonfly_client_core::{
+ use dragonfly_client_core::{
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/health.rs:17]: Please remove the changes on this line or it will not pass the lint test.

**Review by BruceAko — CHANGES_REQUESTED** [review]

File: dragonfly-client/src/grpc/scheduler.rs [hunk 3]
```diff
@@ -13,9 +13,9 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 // use crate::dynconfig::Dynconfig;
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:16]: Please remove this line or add a blank line to the previous line.

**Review by BruceAko — CHANGES_REQUESTED** [review]

File: dragonfly-client/src/grpc/scheduler.rs [hunk 4]
```diff
@@ -29,14 +29,16 @@ use dragonfly_api::scheduler::v2::{
...
 use std::net::{IpAddr, SocketAddr};
 use std::str::FromStr;
 use std::sync::Arc;
+use std::time::Instant;
 use tokio::sync::RwLock;
 use tokio::task::JoinSet;
 use tonic::transport::Channel;
+use tonic_health::pb::health_check_response::ServingStatus;
 use tracing::{error, info, instrument, Instrument};
-
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:39]: Please do not remove this blank line.

**Review by BruceAko — CHANGES_REQUESTED** [review]

**Review by BruceAko — CHANGES_REQUESTED** [review]

File: dragonfly-client/src/grpc/scheduler.rs [hunk 5]
```diff
@@ -459,22 +465,130 @@ impl SchedulerClient {
...
             Err(err) => {
                 error!("connect to {} failed: {}", addr.to_string(), err);
-                if let Err(err) = self.refresh_available_scheduler_addrs().await {
-                    error!("failed to refresh scheduler client: {}", err);
-                };
+                let mut unavailable_scheduler_addrs =
+                    self.unavailable_scheduler_addrs.write().await;
+                unavailable_scheduler_addrs
+                    .entry(addr.addr)
+                    .or_insert_with(std::time::Instant::now);
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:484]: `Instant::now` is enough

File: dragonfly-client/src/grpc/scheduler.rs [hunk 6]
```diff
@@ -459,22 +465,130 @@ impl SchedulerClient {
...
+                        .max_decoding_message_size(usize::MAX)
+                        .max_encoding_message_size(usize::MAX));
+                }
+                Err(err) => {
+                    error!("scheduler {} is not available: {}", scheduler_addr, err);
+                    let mut unavailable_scheduler_addrs =
+                        self.unavailable_scheduler_addrs.write().await;
+                    unavailable_scheduler_addrs
+                        .entry(scheduler_addr)
+                        .or_insert_with(std::time::Instant::now);
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:526]: ditto

File: dragonfly-client/src/grpc/scheduler.rs [hunk 7]
```diff
@@ -580,3 +699,466 @@ impl SchedulerClient {
...
+                        .max_encoding_message_size(usize::MAX));
+                }
+
+                Err(err) => {
+                    error!("connect to {} failed: {}", addr.to_string(), err);
+                    let mut unavailable_scheduler_addrs =
+                        self.unavailable_scheduler_addrs.write().await;
+                    unavailable_scheduler_addrs
+                        .entry(addr.addr)
+                        .or_insert_with(std::time::Instant::now);
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:791]: ditto

File: dragonfly-client/src/grpc/scheduler.rs [hunk 8]
```diff
@@ -580,3 +699,466 @@ impl SchedulerClient {
...
+                            .max_decoding_message_size(usize::MAX)
+                            .max_encoding_message_size(usize::MAX));
+                    }
+                    Err(err) => {
+                        error!("scheduler {} is not available: {}", scheduler_addr, err);
+                        let mut unavailable_scheduler_addrs =
+                            self.unavailable_scheduler_addrs.write().await;
+                        unavailable_scheduler_addrs
+                            .entry(scheduler_addr)
+                            .or_insert_with(std::time::Instant::now);
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:831]: ditto

**Review by BruceAko — CHANGES_REQUESTED** [review]

File: dragonfly-client/src/grpc/scheduler.rs [hunk 9]
```diff
@@ -580,3 +699,466 @@ impl SchedulerClient {
...
+        match scheduler_client
+            .client(task_id, peer_id, &mock_scheduler_client, &mock_dynconfig)
+            .await
+        {
+            Ok(_) => {
+                assert!(false, "Client should not connect successfully.");
+            }
+            Err(e) => match e {
+                Error::ExceededMaxAttempts => {
+                    assert!(true, "Expected ExceededMaxAttempts error.");
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:947]: `assert!(true)` will be optimized out by the compiler. Please use `assert_eq!(e, Error::ExceededMaxAttempts, "Expected ExceededMaxAttempts error.");` instead.

File: dragonfly-client/src/grpc/scheduler.rs [hunk 10]
```diff
@@ -580,3 +699,466 @@ impl SchedulerClient {
...
+            .returning(|| Ok(()));
+
+        let task_id = "task12345";
+        let peer_id = Some("peer12345");
+        match scheduler_client
+            .client(task_id, peer_id, &mock_scheduler_client, &mock_dynconfig)
+            .await
+        {
+            Ok(_) => {
+                assert!(false, "Client should not connect successfully.");
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:943]: `assert!(false, ..)` should probably be replaced by `panic!(..)` or `unreachable!(..)`

File: dragonfly-client/src/grpc/scheduler.rs [hunk 11]
```diff
@@ -580,3 +699,466 @@ impl SchedulerClient {
...
+            }
+            Err(e) => match e {
+                Error::AvailableSchedulersNotFound => {
+                    assert!(true, "Available schedulers not found error.");
+                }
+                _ => {
+                    assert!(false, "Unexpected error type.");
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:1007]: ditto

File: dragonfly-client/src/grpc/scheduler.rs [hunk 12]
```diff
@@ -580,3 +699,466 @@ impl SchedulerClient {
...
+        // Step: 4: Verify that the scheduler is available.
+        match scheduler_client
+            .client(task_id, peer_id, &mock_scheduler_client, &mock_dynconfig)
+            .await
+        {
+            Ok(_) => {
+                assert!(true, "Client should connect successfully.");
+            }
+            Err(_e) => {
+                assert!(false, "Client should connect successfully.")
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:1039]: ditto

File: dragonfly-client/src/grpc/scheduler.rs [hunk 13]
```diff
@@ -580,3 +699,466 @@ impl SchedulerClient {
...
+                    {
+                        Ok(_) => {
+                            assert!(false, "Client should not connect successfully.");
+                        }
+                        Err(e) => match e {
+                            Error::AvailableSchedulersNotFound => {
+                                assert!(true, "Available schedulers not found error.");
+                            }
+                            _ => {
+                                assert!(false, "Unexpected error type.");
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:1112]: ditto

File: dragonfly-client/src/grpc/scheduler.rs [hunk 14]
```diff
@@ -580,3 +699,466 @@ impl SchedulerClient {
...
+                            &mock_scheduler_client,
+                            &mock_dynconfig,
+                        )
+                        .await
+                    {
+                        Ok(_) => {
+                            assert!(true, "Client should connect successfully.");
+                        }
+                        Err(_e) => {
+                            assert!(false, "Client should connect successfully.")
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:1151]: ditto

File: dragonfly-client/src/grpc/scheduler.rs [hunk 15]
```diff
@@ -580,3 +699,466 @@ impl SchedulerClient {
...
+                        }
+                    }
+                })
+            })
+            .collect();
+
+        for handle in handles {
+            if let Err(_e) = handle.await {
+                assert!(false, "One of the tasks panicked.");
```
Comment (BruceAko) [inline dragonfly-client/src/grpc/scheduler.rs:1160]: ditto

**Review by BruceAko — CHANGES_REQUESTED** [review]

**Review by gaius-qi — APPROVED** [review]
LGTM

**Pull request merged**